#!/usr/bin/env Rscript

library("argparse")
parser <- ArgumentParser(description = "Prepare data for microhaplotype analysis")
parser$add_argument(
  "--infile",
  type = "character",
  required = TRUE,
  help = "Path to the outputCIGAR.tsv file generated by the pipeline"
)
parser$add_argument(
    "--rmindel",
    action = "store_true",
    default = FALSE,
    help = "Remove indels from alleles (default: FALSE, keep indels)"
)

parser$add_argument(
  "--outdir",
  type = "character",
  required = TRUE,
  help = "output directory"
)

parser$add_argument(
  "--minimum_nloci",
  type = "integer",
  default = 75,
  help = "Minimum number of loci required for a sample to pass filtering (default: 75)"
)
parser$add_argument(
  "--allele_count_cutoff",
  type = "integer",
  default = 5,
  help = "Minimum read-pair count for each allele (default: 5)"
)
parser$add_argument(
  "--minimum_total_cutoff",
  type = "integer",
  default = 10,
  help = "Minimum total read-pair count per locus per sample (default: 10)"
)

args <- parser$parse_args()

outputCIGAR.file <- args$infile
outdir <- args$outdir
rmindel <- args$rmindel
allele.count.cutoff <- args$allele_count_cutoff
minimum.total.cutoff <- args$minimum_total_cutoff
minimum.nloci <- args$minimum_nloci



if (!dir.exists(outdir)) {
  dir.create(outdir, recursive = TRUE)
}

long.file <-  file.path(outdir, "long_unfiltered.tsv")
mhap.filtered.nloci.per.sample.text  <- file.path(
  outdir,
  "mhap_filtered_nloci.tsv"
)

mhap.filtered.nloci.per.sample.figure <- file.path(
  outdir,
  "mhap_filtered_nloci.pdf"
)

mhap.filtered.subset.file <- file.path(
  outdir,
  "mhap_filtered.tsv"
)

## Load all functions by running everything from line 4-108
outputCIGAR.to.long <- function(outputCIGAR, keep.unused.alleles = FALSE) {
  long <-
    reshape(
      outputCIGAR,
      direction = "long",
      varying = names(outputCIGAR)[-1],
      v.names = "count",
      timevar = "pseudoCIGAR",
      times = names(outputCIGAR)[-1],
      idvar = "sample_id",
      ids = outputCIGAR[["sample"]]
    )
  row.names(long) <- NULL
  
  if (!keep.unused.alleles) {
    long <- long[long[["count"]] > 0, ]
  }
  
  # format data to required columns
  locus.allele <- strsplit(long[["pseudoCIGAR"]], ",")
  long[["locus"]] <- vapply(locus.allele, "[[", character(1), 1)
  long[["allele"]] <- vapply(locus.allele, "[[", character(1), 2)
  
  long[, c("sample_id", "locus", "allele", "count")]
}


rmindel.allele <- function(long) {
  # insertion pattern: {position}I={bases}
  # deletion pattern: {position}D={bases}
  long[["allele"]] <- gsub("([0-9]+)[DI]=([ACGT]+)", "", long[["allele"]])
  
  # default to wild type if all variants were removed
  long[long[["allele"]] %in% "", "allele"] <- "."
  
  # consolidate allele counts after indel removal
  aggregate(count ~ sample_id + locus + allele, long, sum)
}


select.markers <- function(long, keep.marker = "microhaplotype") {
  switch (
    keep.marker,
    microhaplotype = {
      long[!grepl("MIT|DHPS|MDR1", long[["locus"]]), ]
    },
    drugR = {
      long[grepl("DHPS|MDR1", long[["locus"]]), ]
    },
    mitochondria = {
      long[grepl("MIT", long[["locus"]]), ]
    },
    stop("Valid options are 'microhaplotype', 'drugR', 'mitochondria'")
  )
}


allele.count.filter <- function(long, allele.count.cutoff) {
  long[long[["count"]] >= allele.count.cutoff, ]
}


minimum.total.filter <- function(long, minimum.total.cutoff) {
  long.counts.per.locus <- aggregate(count ~ sample_id + locus, long, sum)
  
  pass.filter <-
    long.counts.per.locus[long.counts.per.locus[["count"]] >= minimum.total.cutoff, ]
  
  merge(long, pass.filter[, c("sample_id", "locus")], sort = FALSE)
}


calculate.remaining.nloci <- function(long.filtered) {
  long.filtered.nloci.per.sample <-
    aggregate(locus ~ sample_id, long.filtered, function(x) length(unique(x)))
  
  names(long.filtered.nloci.per.sample)[2] <- "nloci"
  
  long.filtered.nloci.per.sample
}


plot.remaining.nloci <- function(long.filtered.nloci.per.sample, minimum.nloci) {
  plot(
    sort(long.filtered.nloci.per.sample[["nloci"]]),
    main = NA,
    xlab = NA,
    ylab = "Remaining number of locus",
    xaxt = "n"
  )
  
  abline(h = minimum.nloci, col = "red", lty = "dashed")
}


filter.failed.samples <-
  function(long.filtered, long.filtered.nloci.per.sample, minimum.nloci) {
    pass.samples <-
      long.filtered.nloci.per.sample[
        long.filtered.nloci.per.sample[["nloci"]] >= minimum.nloci,
        "sample_id"
      ]
    
    long.filtered[long.filtered[["sample_id"]] %in% pass.samples, ]
  }






outputCIGAR <- read.delim(outputCIGAR.file, check.names = FALSE)
long <- outputCIGAR.to.long(outputCIGAR)

if (rmindel){
  printf("Removing indels from alleles...\n")
  long <- rmindel.allele(long)
}

write.table(long, long.file, quote = FALSE, sep = "\t", row.names = FALSE)



# FIXME: comment any lines below to disable certain filters
mhap <- select.markers(long, keep.marker = "microhaplotype")
mhap.filtered <- allele.count.filter(mhap, allele.count.cutoff)
mhap.filtered <- minimum.total.filter(mhap.filtered, minimum.total.cutoff)
mhap.filtered.nloci.per.sample <- calculate.remaining.nloci(mhap.filtered)



write.table(
  mhap.filtered.nloci.per.sample,
  mhap.filtered.nloci.per.sample.text,
  quote = FALSE,
  sep = "\t",
  row.names = FALSE
)



pdf(file = mhap.filtered.nloci.per.sample.figure)
plot.remaining.nloci(mhap.filtered.nloci.per.sample, minimum.nloci)
dev.off()

mhap.filtered.subset <-
  filter.failed.samples(
    mhap.filtered,
    mhap.filtered.nloci.per.sample,
    minimum.nloci
  )

## STOP and look at the generated text and figure statistics
## Was the remaining locus threshold appropriate for this dataset?


write.table(
  mhap.filtered.subset,
  mhap.filtered.subset.file,
  quote = FALSE,
  sep = "\t",
  row.names = FALSE
)
